freud/diffraction.pyx

```diff
@@ -0,0 +1,288 @@
+        """Calculate the inverse shear matrix from finding the projected box
+        vectors whose area of parallogram is the largest.
+
+        Args:
+            view_orientation ((:math:`4`) :class:`numpy.ndarray`):
+                View orientation.
+            box (:class:`~.box.Box`):
+                Simulation box.
+
+        Returns:
+            (2, 2) :class:`numpy.ndarray`:
+                Inverse shear matrix.
+        """
+        # Rotate the box matrix by the view orientation.
+        box_matrix = rowan.rotate(view_orientation, box.to_matrix())
+
+        # Compute normals for each box face.
+        # The area of the face is the length of the vector.
+        box_face_normals = np.cross(
+            np.roll(box_matrix, 1, axis=-1),
+            np.roll(box_matrix, -1, axis=-1),
+            axis=0)
+
+        # Compute view axis projections.
+        projections = np.abs(box_face_normals.T @ np.array([0., 0., 1.]))
+
+        # Determine the largest projection area along the view axis and use
+        # that face for the projection into 2D.
+        best_projection_axis = np.argmax(projections)
+        secondary_axes = np.array([
+            best_projection_axis + 1, best_projection_axis + 2]) % 3
+
+        # Figure out appropriate shear matrix
+        shear = box_matrix[np.ix_([0, 1], secondary_axes)]
+
+        # Return the inverse shear matrix
+        inv_shear = np.linalg.inv(shear)
+        return inv_shear
+
+    def _transform(self, img, box, inv_shear):
+        """Zoom, shear, and scale diffraction intensities.
+
+        Args:
+            img ((:math:`N`, :math:`N`) :class:`numpy.ndarray`):
+                Array of diffraction intensities.
+            box (:class:`~.box.Box`):
+                Simulation box.
+            inv_shear ((2, 2) :class:`numpy.ndarray`):
+                Inverse shear matrix.
+
+        Returns:
+            (:math:`N`, :math:`N`) :class:`numpy.ndarray`:
+                Transformed array of diffraction intensities.
+        """
+        if img.shape[0] % 2 == 0:
+            roll = img.shape[0] / 2 - 0.5
+        else:
+            roll = img.shape[0] / 2
+        box_matrix = box.to_matrix()
+        ss = np.max(box_matrix) * inv_shear
+
+        shift_matrix = np.array(
+            [[1, 0, -roll],
+             [0, 1, -roll],
+             [0, 0, 1]])
+
+        shear_matrix = np.array(
+            [[ss[1, 0], ss[0, 0], roll],
+             [ss[1, 1], ss[0, 1], roll],
+             [0, 0, 1]])
+
+        zoom_matrix = np.diag((self.zoom, self.zoom, 1))
+
+        # This matrix uses homogeneous coordinates. It is a 3x3 matrix that
+        # transforms 2D points and adds an offset.
+        inverse_transform = np.linalg.inv(
+            zoom_matrix @ shear_matrix @ shift_matrix)
+
+        img = scipy.ndimage.affine_transform(
+            input=img,
+            matrix=inverse_transform,
+            output_shape=(self.grid_size, self.grid_size),
```

Comment:
We should add a separate argument `output_size` or something like that, to de-couple the final image size from the number of bins used for the grid and FFT. Then we could improve the quality of real-time visualization in OVITO.

Reply:
The `k_values` should also be computed based on the `output_size`, not `grid_size`.

Reply:
I've pushed a version that decouples the final img size from grid size. Currently it works well when the `output_size` is smaller than the `grid_size`. For the case where `output_size` is larger than `grid_size`, different zoom value will yield the same output size.